jerry-core/ecma/builtin-objects/ecma-builtin-json.cpp

```diff
@@ -0,0 +1,1815 @@
+{
+  JERRY_ASSERT (reviver_p);
+  JERRY_ASSERT (holder_p);
+  JERRY_ASSERT (name_p);
+
+  ecma_completion_value_t completion_value = ecma_op_object_get (holder_p, name_p);
+
+  if (!ecma_is_completion_value_normal (completion_value))
+  {
+    return completion_value;
+  }
+
+  ecma_value_t value = ecma_copy_value (ecma_get_completion_value_value (completion_value), true);
+  ecma_free_completion_value (completion_value);
+  completion_value = ecma_make_empty_completion_value ();
+
+  if (ecma_is_value_object (value))
+  {
+    ecma_object_t *object_p = ecma_get_object_from_value (value);
+    ecma_property_t *properties_p = ecma_get_property_list (object_p);
+    uint32_t no_properties = 0;
+    ecma_property_t *property_p;
+
+    /*
+     * The following algorithm works with arrays and objects as well.
+     */
+
+    for (property_p = properties_p;
+         property_p != NULL;
+         property_p = ECMA_GET_POINTER (ecma_property_t, property_p->next_property_p))
+    {
+      /*
+       * All properties must be named properties, since we constructed them.
+       */
+      JERRY_ASSERT (property_p->type == ECMA_PROPERTY_NAMEDDATA
+                    || property_p->type == ECMA_PROPERTY_INTERNAL);
+
+      if (property_p->type == ECMA_PROPERTY_NAMEDDATA
+          && property_p->u.named_data_property.enumerable)
+      {
+        no_properties++;
+      }
+    }
+
+    if (no_properties > 0)
+    {
+      size_t buffer_size = sizeof (ecma_string_t*) * (no_properties);
+      ecma_string_t **property_names_p = (ecma_string_t**) mem_heap_alloc_block (buffer_size,
+                                                                                 MEM_HEAP_ALLOC_SHORT_TERM);
+
+      if (property_names_p == NULL)
+      {
+        jerry_fatal (ERR_OUT_OF_MEMORY);
+      }
+
+      uint32_t property_index = no_properties;
+      for (property_p = properties_p;
+           property_p != NULL;
+           property_p = ECMA_GET_POINTER (ecma_property_t, property_p->next_property_p))
+      {
+        if (property_p->type == ECMA_PROPERTY_NAMEDDATA
+            && property_p->u.named_data_property.enumerable)
+        {
+          ecma_string_t *property_name_p = ECMA_GET_NON_NULL_POINTER (ecma_string_t,
+                                                                      property_p->u.named_data_property.name_p);
+         /*
+          * The property list must be in creation order.
+          */
+          property_index--;
+          property_names_p[property_index] = ecma_copy_or_ref_ecma_string (property_name_p);
+        }
+      }
+
+      JERRY_ASSERT (property_index == 0);
+
+      for (property_index = 0; property_index < no_properties; property_index++)
+      {
+        completion_value = ecma_builtin_json_walk (reviver_p,
+                                                   object_p,
+                                                   property_names_p[property_index]);
+        if (!ecma_is_completion_value_normal (completion_value))
```

Comment:
Could we change the function to `ECMA_TRY_CATCH` style with `if (ecma_is_completion_value_empty (ret_value))` as it is performed in other code parts?